adapters/beachfront/beachfront.go

```diff
@@ -197,289 +205,446 @@ func (a *BeachfrontAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *a
-		} else {
-			bfReqs[i].VideoResponseType = "nurl"
+		slot := slot{
+			Id:       appid,
+			Slot:     imp.ID,
+			Bidfloor: beachfrontExt.BidFloor,
 		}
 
-		bfReqs[i].Request = *request
-		var secure int8
-
-		if bfReqs[i].Request.Site != nil && bfReqs[i].Request.Site.Domain == "" && bfReqs[i].Request.Site.Page != "" {
-			bfReqs[i].Request.Site.Domain = getDomain(bfReqs[i].Request.Site.Page)
-
-			secure = isSecure(bfReqs[i].Request.Site.Page)
+		if beachfrontExt.BidFloor <= minBidFloor {
+			slot.Bidfloor = 0
 		}
 
-		if bfReqs[i].Request.App != nil && bfReqs[i].Request.App.Domain == "" && bfReqs[i].Request.App.Bundle != "" {
-			if bfReqs[i].Request.App.Bundle != "" {
-				var chunks = strings.Split(strings.Trim(bfReqs[i].Request.App.Bundle, "_"), ".")
+		for j := 0; j < len(imp.Banner.Format); j++ {
+			slot.Sizes = append(slot.Sizes, size{
+				H: imp.Banner.Format[j].H,
+				W: imp.Banner.Format[j].W,
+			})
+		}
 
-				if len(chunks) > 1 {
-					bfReqs[i].Request.App.Domain =
-						fmt.Sprintf("%s.%s", chunks[len(chunks)-(len(chunks)-1)], chunks[0])
-				}
+		if len(slot.Sizes) == 0 {
+			if imp.Banner.W == nil || imp.Banner.H == nil {
+				errs = append(errs,
+					fmt.Errorf("request.imp[%d].banner has no sizes. Define \"w\" and \"h\", or include \"format\" elements", i))
+				continue
+			} else {
+				slot.Sizes = append(slot.Sizes, size{
+					W: *imp.Banner.W,
+					H: *imp.Banner.H,
+				})
 			}
 
 		}
 
-		if bfReqs[i].Request.Device.DeviceType == 0 {
-			// More fine graned deviceType methods will be added in the future
-			bfReqs[i].Request.Device.DeviceType = fallBackDeviceType(request)
-		}
+		bfr.Slots = append(bfr.Slots, slot)
+	}
 
-		imp := request.Imp[i]
+	return bfr, secure, errs
+}
 
-		imp.Banner = nil
-		imp.Ext = nil
-		imp.Secure = &secure
+func getVideoRequests(request *openrtb.BidRequest, imps []openrtb.Imp, errs []error) ([]openrtb.BidRequest, map[string]*openrtb.BidRequest, []error) {
+	var nurlReqs []openrtb.BidRequest
 
-		if beachfrontExt.BidFloor <= minBidFloor {
-			imp.BidFloor = 0
-		} else {
-			imp.BidFloor = beachfrontExt.BidFloor
+	// The pointers below will each point to a deep copy of the request argument, cleaned
+	// up and stripped of the original imps, which are replaced with deep copies in the provided
+	// imps slice. Everything is fresh and new.
+	var admMap = map[string]*openrtb.BidRequest{}
+
+	for i := 0; i < len(imps); i++ {
+		ext, err := parseBeachfrontExtension(imps[i], openrtb_ext.BidTypeVideo)
+
+		if err != nil {
+			errs = append(errs, errors.New(fmt.Sprintf("%s (on video imp id: %s)", err, imps[i].ID)))
+			continue
 		}
 
-		if imp.Video.H == 0 && imp.Video.W == 0 {
-			imp.Video.W = DefaultVideoWidth
-			imp.Video.H = DefaultVideoHeight
+		if ext.VideoResponseType == "nurl" || ext.VideoResponseType == "both" {
+			r := deepCopyRequest(request)
+			r.Imp = []openrtb.Imp{imps[i]}
+
+			nurlReqs = append(nurlReqs, prepVideoRequest(r))
```

Comment:
Sounds like you are making some progress. I am not sure if this will help the problem, but it should make things more efficient in any case. Instead of running `prepVideoRequest()` for each request you generate, run it once on your deep copy of the original request, then do a shallow copy of the prepped request here in order to replace `request.imp` in this instance of your generated requests. If it is somehow `prepVideoRequest()` that is somehow stepping on its own toes, then running it only once will keep that from happening. In any case, it doesn't look like there is any dependence on the `imp` in prep, which means it should be doing the same exact thing every time you execute it. This will at least make your adapter more efficient even if it doesn't fix the race issue.

Reply:
Weirdly, I don't seem to be able to comment on here any more. Unless you see this comment, in which case I now can.

Reply:
It was Brave browser, blocking that which it should not block ;)